feat: Add TOML support to the built-in content loaders

aditsachde opened the pull request:
## Changes

Add TOML support to the built-in content loaders.

A bit over a year ago, there was a PR (https://github.com/withastro/astro/pull/11004) which added TOML support to Astro, but it was closed in favor of a community loader. However, in Astro 5.2, support was added for TOML Markdown frontmatter (including within content collections). 

I think that following the addition of that feature, it makes sense to directly support TOML files within content collections for consistency. I wanted to move some data I had in markdown frontmatter to just plain TOML files and it feels odd to have support for the format in one place but not others.

I've added `smol-toml` as a direct dependency of astro, as this was already a transitive dependency through the markdown package.

## Testing

This change was tested by extending the current test fixtures to include TOML files.

## Docs

This will need documentation! I am happy to update the docs with a bit of guidance. 

I think at minimum, we'd want to update the following link to mention that both the glob and file loaders now natively support toml.
https://docs.astro.build/en/reference/content-loader-reference/#built-in-loaders


In addition, the example about the custom parser function uses the toml format as an example. We'd probably want to update this to a CSV or another format, as if this change is merged, we will have native toml support and the example wouldn't make much sense.
https://docs.astro.build/en/guides/content-collections/#parser-function

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
/cc @withastro/maintainers-docs for feedback!

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

ascorbic commented:
@aditsachde changing the custom loader test to use a different filetype would be best I think. You can then add a test with multiple files so that glob works properly. Thanks!

aditsachde commented:
I've added toml and yaml tests for the glob loader and updated the file loader tests to use the native toml functionality instead of the custom parser option of the file content collection. I've added a CSV loader test to make sure we continue to exercise the custom parser option of the file content collection.

aditsachde commented:
Hi! I wanted to follow up on the documentation updates needed for this change. It looks like this change missed 5.11, but it'd be nice to get this in as part of 5.12 or 5.13.

delucis commented:
Hey @aditsachde — thanks for the PR!

An accompanying docs PR would be super helpful. I think your suggestions sound good. There are a few places (e.g. in the loader [docs](https://docs.astro.build/en/guides/content-collections/#built-in-loaders) and [reference](https://docs.astro.build/en/reference/content-loader-reference/#file-loader)) where we mention “JSON and YAML support” that will need updating to mention TOML as well. It’s worth doing a sweep of different pages to make sure we don’t miss a spot.

As you spotted, the `parser` docs show how to add TOML support which wouldn’t make much sense anymore, so we should remove that. The example does already show adding CSV support, so I think just leaving that as a single example is probably enough.

Let us know if you have any questions. We have a [dedicated guide to docs contributions](https://contribute.docs.astro.build/) in case it’s helpful.

ascorbic commented:
I've added this to the 5.12 milestone which is released this Thursday, the 17th, so ideally we'd need these docs by the end of Wednesday. Let us know if you won't be able to have them done by then, or need somebody else to handle them

sarah11918 commented on .changeset/icy-keys-tickle.md:
```diff
@@ -0,0 +1,10 @@
+
+Add TOML support to the built-in content loaders.
+
+In Astro 5.2, Markdown gained support for TOML frontmatter, including for use within content collections.
+
+This changeset extends that functionality to directly support TOML files in content collections.
```
``````suggestion
In Astro 5.2, Astro added support for using TOML frontmatter in Markdown files instead of YAML. However, if you wanted to use TOML files as local content collection entries themselves, you needed to import your own parser and pass it to Astro's loader.

Astro 5.12 now directly supports loading data from TOML files in content collections in both the `glob()` and the `file()` loaders.

If you had added your own TOML content parser,  you can now remove it as this functionality is now included:

```diff
// src/content.config.ts

import { defineCollection } from "astro:content";
import { file } from "astro/loaders";
- import { parse as parseToml } from "toml";

const dogs = defineCollection({
-  loader: file("src/data/dogs.toml", { parser: (text) => parseToml(text).dogs }),
+ loader: file()
  schema: /* ... */
})
```

See Astro's [content collections guide](https://docs.astro.build/en/guides/content-collections/#built-in-loaders) for more information on using the built-in content loaders.
``````
(I am notoriously bad at predicting what formatting will look like, but the above is the general idea!)

Since someone might possibly have a project that would benefit from updating, it's nice to show some kind of example!

aditsachde replied:
Thanks! I've updated the changeset, pulling largely from your example. However, I also added a note about some specific quirks with toml and how that interacts with the file loader.

sarah11918 reviewed (commented):
Great addition to the built in loaders! I did check, and Chris's comment re: a docs PR should be sufficient.

Just some thoughts here re: the changeset for your consideration!

sarah11918 commented:
Thanks for letting us know @aditsachde ! We'll get someone to handle the docs for this feature so that it can release on Thursday!

sarah11918 commented:
(Please check the accompanying docs PR, too! https://github.com/withastro/docs/pull/12029 )

sarah11918 commented on packages/astro/src/core/errors/errors-data.ts:
```diff
@@ -1855,7 +1855,7 @@ export const UnsupportedConfigTransformError = {
  *  - [Passing a `parser` to the `file` loader](https://docs.astro.build/en/guides/content-collections/#parser-function)
  * @description
- * The `file` loader can’t determine which parser to use. Please provide a custom parser (e.g. `toml.parse` or `csv-parse`) to create a collection from your file type.
+ * The `file` loader can’t determine which parser to use. Please provide a custom parser (e.g. or `csv-parse`) to create a collection from your file type.
```
```suggestion
 * The `file` loader can’t determine which parser to use. Please provide a custom parser (e.g. `csv-parse`) to create a collection from your file type.
```
